### ISSUE TITLE
Support adding provisioned machines to the juju model

### ISSUE THREAD

**Issue by aieri** [juju/terraform-provider-juju#103]
Juju has traditionally lacked support for provisioning machines, delegating that task to the resource provider. As an example, in a baremetal environment you may be storing your hardware configuration in maas (i.e. network and storage configs) and your application configuration in a juju bundle; although juju knows how to `add-machine` and grab a physical node from the maas pool, it is generally blind to how that node may have been set up. Operators may carefully assign maas nodes to tags that will be leveraged on the juju side as application constraints, but the two worlds generally remain separate.

In the long run, it would be nice for juju to be able to drive hardware configurations so that users may specify their entire deployment in a single cohesive place, but for now perhaps terraform could be used as an interim solution: since several terraform providers already exist (even a [maas one](https://maas.io/docs/maas-terraform-provider)), and since the terraform provider for juju allows terraform to drive a juju model, if there was support for adding an already terraform-provisioned machine to a juju model - ideally even retaining the correct "type" based on the providing backend - one could already define an entire environment in a single place, perhaps even reducing or eliminating the need for custom scripts to automate complex deployments.

In other words, given an existing juju controller using the maas resource provider, I'm imagining being able to do the following:
1. provision a machine in maas via terraform, using the [terraform provider for MAAS](https://github.com/maas/terraform-provider-maas). If I understand correctly, the machine will at this point exist in a deployed state
2. use the terraform provider for juju to add the provisioned machine to a model as a native maas-provided machine (not using the manual provider) without redeploying it (the machine is not in ready state)
3. continue using the terraform provider for juju to deploy applications onto the newly-added machine

Of these three steps, I believe the only one needing to be implemented is 2.

**Comment by jnsgruk** [comment]
Yeh, I think there is a use case here in something like OSM too, where machines often need to be deployed from a gold image or similar.

This would allow users to deploy and provision a machine however they like, then assuming prereqs are in place, use juju to deploy charms to it.

**Comment by juanmanuel-tirado** [comment]
This is an interesting use case. It's true that we have focused this terraform provider to manage "pure" Juju entities such as models, applications or integrations ignoring underlying machinery. I think this is a reasonable request.

**Comment by jnsgruk** [comment]
It's essentially contorting the idea behind `add-machine` into a Terraform resource type, perhaps `juju_machine` or similar?

**Comment by juanmanuel-tirado** [comment]
Yes something like that would follow the current schema.

**Comment by VariableDeclared** [comment]
I believe this could be achieved with something like: 
```
resource "juju_application" "containerd" {
  name = "containerd"

  model = juju_model.k8s-tf.name

  charm {
    name = "containerd"
  }
  units = 0
  machines = [azurerm_linux_virual_machine.private_ip[0], another machine, etc]

}
```

I am also wishing for such a feature, if my use case above is covered, then this bug covers it :)
```

**Comment by VariableDeclared** [comment]
Building on the last comment, I think what is missing is a juju_controller resource, 
```
resource "juju_controller" "terraform_controller" {

       machines = [azurerm_linux_virtual_machine.my_controller.private_ip[0]]
       .... charm etc
}
```

it would be super cool for the juju controller resource to exist, and to be able to map machines

right now I demo this on azure, the azure resources are created in terraform, but then I have to 'glue' the rest, and deploy k8s with terraform

**Comment by juanmanuel-tirado** [comment]
We think this would definitively improve the whole terraform+juju experience and it is something we have in the queue. Please stay tuned :)

**Comment by juanmanuel-tirado** [comment]
Machines provisioned with Juju are now supported (see #165). This is not fully covering all the use cases but it may be helpful until we have a more complete solution.
